ksqldb-engine/src/main/java/io/confluent/ksql/services/DefaultConnectClientFactory.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.services;
+
+import io.confluent.ksql.services.ConnectClient.ConnectClientFactory;
+import io.confluent.ksql.util.FileWatcher;
+import io.confluent.ksql.util.KsqlConfig;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Paths;
+import java.util.Base64;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import org.apache.kafka.common.config.ConfigException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DefaultConnectClientFactory implements ConnectClientFactory {
+
+  private static final Logger log = LoggerFactory.getLogger(DefaultConnectClientFactory.class);
+
+  private final KsqlConfig ksqlConfig;
+  private Optional<String> connectAuthHeader;
+  private FileWatcher credentialsFileWatcher;
+
+  public DefaultConnectClientFactory(
+      final KsqlConfig ksqlConfig
+  ) {
+    this.ksqlConfig = Objects.requireNonNull(ksqlConfig, "ksqlConfig");
+  }
+
+  @Override
+  public DefaultConnectClient get(final Optional<String> ksqlAuthHeader) {
+    if (connectAuthHeader == null) {
+      connectAuthHeader = buildAuthHeader();
+    }
+
+    return new DefaultConnectClient(
+        ksqlConfig.getString(KsqlConfig.CONNECT_URL_PROPERTY),
+        // if no explicit header specified, then forward incoming request header
+        connectAuthHeader.isPresent() ? connectAuthHeader : ksqlAuthHeader
+    );
+  }
+
+  @Override
+  public void close() {
+    if (credentialsFileWatcher != null) {
+      credentialsFileWatcher.shutdown();
+    }
+  }
+
+  private Optional<String> buildAuthHeader() {
+    // custom basic auth credentials
+    if (ksqlConfig.getString(KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_SOURCE_PROPERTY)
+        .equalsIgnoreCase(KsqlConfig.BASIC_AUTH_CREDENTIALS_SOURCE_FILE)) {
+      final String credentialsFile =
+          ksqlConfig.getString(KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_FILE_PROPERTY);
+      final boolean failOnUnreadableCreds =
+          ksqlConfig.getBoolean(KsqlConfig.CONNECT_BASIC_AUTH_FAIL_ON_UNREADABLE_CREDENTIALS);
+
+      if (ksqlConfig.getBoolean(KsqlConfig.CONNECT_BASIC_AUTH_CREDENTIALS_RELOAD_PROPERTY)) {
+        startBasicAuthFileWatcher(credentialsFile, failOnUnreadableCreds);
+      }
+
+      return buildBasicAuthHeader(credentialsFile, failOnUnreadableCreds);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  private void startBasicAuthFileWatcher(
+      final String filePath,
+      final boolean failOnUnreadableCreds
+  ) {
+    try {
+      credentialsFileWatcher = new FileWatcher(Paths.get(filePath), () -> {
```

Comment:
AFAIU the file watcher does not notify on the file deletion, so if the credentials file is removed, we'll still use the most recent header. Is this OK for our purposes? An alternative would be to erase the header on file deletion. I'd probably add a short comment addressing this scenario in any case.

Reply:
Yes, this is fine for our purposes. I considered updating the FileWatcher implementation to watch for this but it seems preferable to continue using outdated credentials/TLS certs rather than no credentials/certs in the event of file deletion. I will add some comments as you suggested.